sqlmesh/core/dialect.py

```diff
@@ -755,12 +755,19 @@ def select_from_values_for_batch_range(
         exp.alias_(exp.cast(column, to=kind), column, copy=False)
         for column, kind in columns_to_types.items()
     ]
-    values_exp = exp.values(
-        [tuple(transform_values(v, columns_to_types)) for v in values[batch_start:batch_end]],
-        alias=alias,
-        columns=columns_to_types,
-    )
-    return exp.select(*casted_columns).from_(values_exp, copy=False)
+
+    if not values:
+        # Ensures we don't generate an empty VALUES clause & forces a zero-row output
+        where = exp.false()
+        expressions = [tuple(exp.cast(exp.null(), to=kind) for _, kind in columns_to_types.items())]
```

Comment:
FYI I added casts for the `NULL`s inside of the `VALUES` clause too because of a BQ issue I came across:

```
SELECT CAST(json_col AS JSON) AS json_col
FROM UNNEST([STRUCT(NULL AS id, NULL AS ds, NULL AS json_col)])
WHERE FALSE
```

Apparently this fails with "Invalid cast from INT64 to JSON at [1:13]"